### ISSUE TITLE
Modify RUNPATH of Binaries Before Packaging

### ISSUE THREAD

**Issue by danyeaw** [pyinstaller/pyinstaller#6444]
I am trying to create a statically linked AppImage of a PyGObject app using PyInstaller to increase compatibility with older versions of GLIBC. The AppImage works good before trying to statically link it using [staticx](https://github.com/JonathonReinhart/staticx). However, I am not able to link a few of the binaries because the `DT_RUNPATH` is set to an absolute path.

```
staticx: Unsupported PyInstaller input

One or more libraries included in the PyInstaller archive uses unsupported RPATH/RUNPATH tags:

  /tmp/staticx-pyi-1wbu1ila/gio_modules/libgioremote-volume-monitor.so: DT_RUNPATH='/usr/lib64/gvfs'
  /tmp/staticx-pyi-1wbu1ila/gio_modules/libgvfsdbus.so: DT_RUNPATH='/usr/lib64/gvfs'
  /tmp/staticx-pyi-1wbu1ila/libjpeg.so.8: DT_RUNPATH='/usr/lib64'
```

It would be nice if there was a way to set the runpath using something like `chrpath -r @ORIGIN/.. libgiremote-volume-monitor.so` while freezing the app so that it can be statically linked.

**Is your feature request related to a problem? Please describe.**
I would like a way to alter the RUNPATH of binaries prior to them getting added in to a onefile executable.

**Describe the solution you'd like**
Add an option to set the RUNPATH of certain binaries.

**Describe alternatives you've considered**
- Using onedir mode and then try to use [patchelf](https://github.com/NixOS/patchelf) to relink the binaries and executables to the older libc version.
- Using [appimage-builder](https://github.com/AppImageCrafters/appimage-builder) to package the app
- Try to compile GTK from source to improve RPATH compatibility

**Additional context**

**Comment by bwoodsend** [comment]
Isn't the whole idea behind AppImage that it bundles glibc and the like so that use of staticx is redundant?

**Comment by rokm** [comment]
Ah yes, runpaths. Because not a day goes by where I wish that our linux assembly pipeline was a bit more like the macOS one, with its library path rewriting magic...

But on a serious note, it would be a good idea to detect absolute rpath/runpath and attempt to mitigate them, because in addition to `staticx` issues, they might also lead to system libraries being loaded instead of the bundled ones.

**Comment by rokm** [comment]
And FWIW, I don't think changing `DT_RUNPATH` would solve this particular case, because `staticx` does not like RUNPATH in general (whereas it tolerates relative RPATHs) - see [here](https://github.com/JonathonReinhart/staticx/blob/819d8eafecbaab3646f70dfb1e3e19f6bbc017f8/staticx/api.py#L238-L268). So we'd need to change all RUNPATHs into RPATHs, and ensure none of RPATHs are absolute...

**Comment by danyeaw** [comment]
> Isn't the whole idea behind AppImage that it bundles glibc and the like so that use of staticx is redundant?

For the user of an AppImage, yes. Unfortunately, for the creator, you need to either build your app on Ubuntu 16.04 so that an old GLIBC version is used, or statically compile it, or modify the ELFs to point the executable to an older libc.

Trying to build a new version of GTK and the rest of the dependencies on a really old distro version isn't very straightforward.

**Comment by danyeaw** [comment]
@rokm It sounds like you agree that trying to handle rpath/runpath better in PyInstaller would be good. Let me put a smaller example together, and then we could discuss how best to go about it.

**Comment by danyeaw** [comment]
I am closing this issue, but I thought I would document what I found if others come across it.

I think this is currently working because RUNPATH is lower priority than LD_LIBRARY_PATH. RPATH takes the highest precedence. So although we could add support for clearing the RUNPATH, in practice it probably isn't that important.

I was able to get what I needed by using JHBuild to build GTK on Ubuntu 18.04, instead of taking the staticx approach.
